In [1]:
from dbfread import DBF
import pandas as pd
from pandas import DataFrame

## Load 10 years of accident data, from 2007 to 2016

In [13]:
dbf07= DBF('accident/accident2007.dbf')
dbf08= DBF('accident/accident2008.dbf')
dbf09= DBF('accident/accident2009.dbf')
dbf10= DBF('accident/accident2010.dbf')
dbf11 = DBF('accident/accident2011.dbf')
dbf12 = DBF('accident/accident2012.dbf')
dbf13 = DBF('accident/accident2013.dbf')
dbf14 = DBF('accident/accident2014.dbf')
dbf15 = DBF('accident/accident2015.dbf')
dbf16 = DBF('accident/accident2016.dbf')
accidents07 = DataFrame(iter(dbf07))
accidents08 = DataFrame(iter(dbf08))
accidents09 = DataFrame(iter(dbf09))
accidents10 = DataFrame(iter(dbf10))
accidents11 = DataFrame(iter(dbf11))
accidents12 = DataFrame(iter(dbf12))
accidents13 = DataFrame(iter(dbf13))
accidents14 = DataFrame(iter(dbf14))
accidents15 = DataFrame(iter(dbf15))
accidents16 = DataFrame(iter(dbf16))


First, we want to combine accidents10 ~ accidents16 to one dataframe. Since not all of the accident data downloaded from the U.S. Department of Transportation have the same features, by using the `jion:inner` option in `pd.concat` function, we can get the intersection of features.

In [14]:
# rename column name in frame07 so that columns names are the same with other frames
accidents07.rename(columns={'latitude': 'LATITUDE', 'longitud': 'LONGITUD'}, inplace=True)

In [15]:
# take a look inside how the accident data file looks like
allaccidents = pd.concat([accidents07,accidents08,accidents09,accidents10,accidents11,accidents12,accidents13,accidents14,accidents15,accidents16], axis=0,join='inner')
pd.set_option('display.max_columns', 100)

In [ ]:
#import datetime
#months = {i :datetime.datetime(2000,i,1).strftime("%B") for i in range(1, 13)}
#type(datetime.datetime(2000,1,1).strftime("%b"))
#accidents["MONTH"] = accidents["MONTH"].map(months)
#accidents["MONTH"]
#accidents.head()
#df.sort(columns=[...])

In [16]:
#combine all vehicle file
import warnings
warnings.filterwarnings('ignore')
accidents = allaccidents[['YEAR','ST_CASE','STATE','VE_TOTAL','PERSONS','FATALS','MONTH','DAY_WEEK','HOUR','NHS','LATITUDE','LONGITUD','MAN_COLL','LGT_COND','WEATHER','ARR_HOUR','ARR_MIN','CF1','DRUNK_DR']]
accidents.rename(columns={'ST_CASE':'CASE_NUM','VE_TOTAL':'NUM_VEHICLE','NHS': 'HIGHWAY', 'MAN_COLL': 'COLLISION_TYPE','LGT_COND':'LIGHT_CONDITION','CF1':'CRASH_FACTOR','DRUNK_DR':'DRUNK_DRIVE'}, inplace=True)
accidents['MONTH'] = accidents['MONTH'].map({1.0:'January', 2.0:'February', 3.0: 'March', 4.0:'April', 5.0:'May', 6.0:'June', 7.0:'July', 8.0:'August',9.0: 'September', 10.0:'October', 11.0:'November', 12.0:'December'})
accidents['DAY_WEEK']= accidents['DAY_WEEK'].map({1.0:'Sunday',2.0:'Monday', 3.0:'Tuesday', 4.0: 'Wednesday', 5.0:'Thursday', 6.0:'Friday', 7.0:'Saturday'})
accidents['HIGHWAY'] = accidents['HIGHWAY'].map({1.0:'On',0.0:'Off',9.0:'Unknow'})
accidents['COLLISION_TYPE'] = accidents['COLLISION_TYPE'].map({0.0:'Not Collision',1.0:'Rear-End',2.0:'Head-On',3.0:'Rear-to-Rear',4.0:'Angle',5.0:'Sideswipe, Same Direction',6.0:'Sideswipe, Opposite Direction',7.0:'Sideswipe, Unknown Direction',9.0:'Unknown'})
accidents['LIGHT_CONDITION'] = accidents['LIGHT_CONDITION'].map({1.0:'Daylight',2.0:'Dark' ,3.0:'Dark',5.0:'Dusk',6.0:'Dark',4.0:'Dawn', 7.0:'Other',8.0 :'Not Report', 9.0:'Not Report'})
# accidents['WEATHER'] = accidents['WEATHER'].map({0.0:'Normal',1.0:'Clear',2.0:'Rain',3.0
accidents.head()

YEAR  CASE_NUM  STATE  NUM_VEHICLE  PERSONS  FATALS    MONTH  DAY_WEEK  \
0  2007.0   10001.0    1.0          1.0      1.0     1.0  January   Tuesday   
1  2007.0   10002.0    1.0          3.0      4.0     2.0  January   Tuesday   
2  2007.0   10003.0    1.0          2.0      5.0     1.0  January    Sunday   
3  2007.0   10004.0    1.0          2.0      2.0     1.0  January    Monday   
4  2007.0   10005.0    1.0          2.0      5.0     1.0  January  Saturday   

   HOUR HIGHWAY   LATITUDE   LONGITUD                COLLISION_TYPE  \
0  23.0     Off  33.455839 -87.017928                 Not Collision   
1  13.0     Off  34.160597 -85.678075     Sideswipe, Same Direction   
2  12.0     Off  33.975717 -86.496347  Sideswipe, Unknown Direction   
3  15.0      On  33.853258 -85.915622                         Angle   
4  19.0     Off  32.488833 -86.408936     Sideswipe, Same Direction   

  LIGHT_CONDITION  WEATHER  ARR_HOUR  ARR_MIN  CRASH_FACTOR  DRUNK_DRIVE  
0            Dark      1.0      23.0     20.0           0.0          0.0  
1        Daylight      1.0      13.0     11.0           0.0          0.0  
2        Daylight      1.0      12.0     18.0           0.0          0.0  
3        Daylight      1.0      15.0      5.0           0.0          0.0  
4            Dark      1.0      19.0     41.0           0.0          0.0

In [17]:
# allaccidents['ST_CASE'] = 'n' + df['ST_CASE']

In [18]:
accidents['STATE']=accidents['STATE'].astype(int)
accidents['CASE_NUM']=accidents['CASE_NUM'].astype(int)
accidents['YEAR']=accidents['YEAR'].astype(int)

In [19]:
accidents.index = list(accidents['YEAR'].astype(str) + accidents['CASE_NUM'].astype(str))
accidents.head()

pandas.core.indexes.base.Index

In [21]:
accidents.shape

(320874, 19)

The allaccidents table recorded 320874 accidents from 2010-2016, and it has 42 features. Here are the meaning of some of the features according to the `FARS Analytical User’s Manual`.

### Explaination of variables
*VE_TOTAL*: Number of Vehicle in crash <br/>
*VE_FORMS*: Number of Motor Vehicles in Transport (MVIT) <br/>
*PED*: Number of Persons Not in Motor Vehicles <br/>
*NHS*: National Highway System<br/>
*ROUTE*: Route Signing <br/>
*SP_JUR*: Special Jurisdiction <br/>
*HARM_EV*: First Harmful Event<br/>
*TWAY_ID , TWAY_ID2* : Trafficway Identifier <br/>
*MILEPT*: Milepoint <br/>
*SP_JUR*: Special Jurisdiction<br/>
*HARM_EV*: injury or damage producing First Harmful Event <br/> 
*MAN_COLL*:Manner of Collision  <br/> 
*RELJCT1, RELJCT2*: Relation to Junction- Within Interchange Area, Specific Location. <br/>
*TYP_INT*: Type of Intersection <br/>
*REL_ROAD*: Relation to Trafficway <br/>
*LGT_COND*: Light Condition<br/> 
*NOT_HOUR,MIN*: Min, Hour of Notification <br/>
*ARR_HOUR,MIN*: Hour, Min arrival at scene <br/>
*HOSP_HR,MIN*: Hour, Min arrival at hospital <br/>
*CF1, CF2, CF3*:Related Factors- Crash Level, factors related to the crash <br/>
*FATALS*: Fatalities<br/>
*DRUNK_DR*: Number of Drinking Drivers<br/> 
*RAIL*: Rail Grade Crossing Identifier<br/>

For more detailed information, please refer to `FARS Analytical User’s Manual`.

## Select variables and rename variables

Observed from the table above, some of the variables in the table are not very readable. Therefore, in order to make it easier to understand the variables,we renamed some of the variables according to `FARS Analytical User’s Manual`  downloaded from the  `U.S. Department of Transportation`  website.

### Load vehicle data file which contains mortality rate

We also want to study the mortality rate of fatal accidents. The data element “Fatalities in Vehicle” in the Vehicle data file from the `U.S. Department of Transportation` website provides the number of deaths in a vehicle.

In [22]:
vdbf07= DBF('vehicle_deaths/vehicle2007.dbf')
vdbf08= DBF('vehicle_deaths/vehicle2008.dbf')
vdbf09= DBF('vehicle_deaths/vehicle2009.dbf')
vdbf10= DBF('vehicle_deaths/vehicle2010.dbf')
vdbf11= DBF('vehicle_deaths/vehicle2011.dbf')
vdbf12= DBF('vehicle_deaths/vehicle2012.dbf')
vdbf13= DBF('vehicle_deaths/vehicle2013.dbf')
vdbf14= DBF('vehicle_deaths/vehicle2014.dbf')
# vdbf15= DBF('vehicle_deaths/vehicle2015.csv')
vdbf16= DBF('vehicle_deaths/vehicle2016.dbf')
vehicle07 = DataFrame(iter(vdbf07))
vehicle08 = DataFrame(iter(vdbf08))
vehicle09 = DataFrame(iter(vdbf09))
vehicle10 = DataFrame(iter(vdbf10))
vehicle11 = DataFrame(iter(vdbf11))
vehicle12 = DataFrame(iter(vdbf12))
vehicle13 = DataFrame(iter(vdbf13))
vehicle14 = DataFrame(iter(vdbf14))
# vehicle15 = pd.read_csv('vehicle_deaths/vehicle2015.csv')
vehicle16 = DataFrame(iter(vdbf16))

In [44]:
vehicle07['YEAR']=2007
vehicle08['YEAR']=2008
vehicle09['YEAR']=2009
vehicle10['YEAR']=2010
vehicle11['YEAR']=2011
vehicle12['YEAR']=2012
vehicle13['YEAR']=2013
vehicle14['YEAR']=2014
# vehicle15['YEAR']='2015.0'
vehicle16['YEAR']=2016

In [45]:
allvehicles=pd.concat([vehicle07,vehicle08,vehicle09,vehicle10,vehicle11,vehicle12,vehicle13,vehicle14,vehicle16], axis=0,join='outer')
vehicles = allvehicles[['STATE','YEAR','ST_CASE','HIT_RUN','TRAV_SP','ROLLOVER','FIRE_EXP','SPEEDREL','DEATHS']]
vehicles.rename(columns={'ST_CASE':'CASE_NUM','TRAV_SP':'SPEED','FIRE_EXP': 'FIRE','SPEEDREL':'SPEEDING'}, inplace=True)
vehicles['STATE']=vehicles['STATE'].astype(int)
vehicles['CASE_NUM']=vehicles['CASE_NUM'].astype(int)
vehicles['YEAR']=vehicles['YEAR'].astype(int)
vehicles.index = list(vehicles['YEAR'].astype(str) + vehicles['CASE_NUM'].astype(str))
vehicles.head()

STATE  YEAR  CASE_NUM  HIT_RUN  SPEED  ROLLOVER  FIRE  SPEEDING  \
200710001      1  2007     10001      0.0   65.0       0.0   0.0       NaN   
200710002      1  2007     10002      0.0   10.0       0.0   0.0       NaN   
200710002      1  2007     10002      0.0   55.0       0.0   0.0       NaN   
200710002      1  2007     10002      0.0   55.0       0.0   0.0       NaN   
200710003      1  2007     10003      1.0   55.0       0.0   0.0       NaN   

           DEATHS  
200710001     1.0  
200710002     0.0  
200710002     2.0  
200710002     0.0  
200710003     0.0

In [51]:
all = pd.merge(vehicles, accidents, left_index=True, right_index=True, how='inner',on=('STATE', 'YEAR','CASE_NUM'))
all.index=(all.index).astype(int)
all.sort_index()
all.head()

STATE  YEAR  CASE_NUM  HIT_RUN  SPEED  ROLLOVER  FIRE  SPEEDING  \
2007100001     10  2007    100001      0.0   99.0       0.0   0.0       NaN   
2007100002     10  2007    100002      0.0   50.0       0.0   0.0       NaN   
2007100002     10  2007    100002      0.0   50.0       0.0   0.0       NaN   
2007100003     10  2007    100003      0.0   45.0       0.0   0.0       NaN   
2007100004     10  2007    100004      2.0   30.0       0.0   0.0       NaN   

            DEATHS  NUM_VEHICLE  PERSONS  FATALS    MONTH   DAY_WEEK  HOUR  \
2007100001     1.0          1.0      1.0     1.0  January    Tuesday  20.0   
2007100002     0.0          2.0      3.0     1.0  January  Wednesday  18.0   
2007100002     0.0          2.0      3.0     1.0  January  Wednesday  18.0   
2007100003     0.0          1.0      2.0     1.0  January   Thursday  10.0   
2007100004     0.0          1.0      2.0     1.0  January     Sunday  11.0   

           HIGHWAY   LATITUDE   LONGITUD COLLISION_TYPE LIGHT_CONDITION  \
2007100001     Off  38.645422 -75.210097  Not Collision            Dark   
2007100002     Off  38.973942 -75.452531  Not Collision            Dark   
2007100002     Off  38.973942 -75.452531  Not Collision            Dark   
2007100003     Off  39.733614 -75.638883  Not Collision        Daylight   
2007100004     Off  39.734878 -75.574147  Not Collision        Daylight   

            WEATHER  ARR_HOUR  ARR_MIN  CRASH_FACTOR  DRUNK_DRIVE  
2007100001      1.0      21.0      2.0           0.0          0.0  
2007100002      1.0      18.0     29.0           0.0          0.0  
2007100002      1.0      18.0     29.0           0.0          0.0  
2007100003      1.0      10.0      5.0           0.0          0.0  
2007100004      1.0      11.0     57.0           0.0          0.0

In [58]:
#the total accidents number each year, analysis the difference between every year
year_acci=all[['YEAR','CASE_NUM']].groupby('YEAR').count()
month_acci=all[['MONTH','CASE_NUM']].groupby('MONTH').count()
day_acci=all[['DAY_WEEK','CASE_NUM']].groupby('DAY_WEEK').count()
hour_acci=all[['HOUR','CASE_NUM']].groupby('HOUR').count()
time_relate_acci=pd.DataFrame({'year_acci':year_acci,'month_acci':month_acci,'day_acci':day_acci,'hour_acci':hour_acci})
time_relate_acci.head()

ValueError: If using all scalar values, you must pass an index

In [ ]:
year_acci.index

In [ ]:
with sns.plotting_context("poster"):
    g = sns.FacetGrid(data=accidents, col="continent", col_wrap=4)
    g.map(sns.boxplot, 'variable',"value")
    g.set_axis_labels("drinks", "consumption value")